README.md
# react-keeptrack
this is react sample integrated with api


history.md
Contribution: 2014-12-08 18:54

Contribution: 2014-12-08 18:55

Contribution: 2014-12-08 18:56

Contribution: 2014-12-08 18:57

Contribution: 2014-12-09 18:54

Contribution: 2014-12-09 18:55

Contribution: 2014-12-09 18:56

Contribution: 2014-12-09 18:57

Contribution: 2014-12-10 18:54

Contribution: 2014-12-10 18:55

Contribution: 2014-12-10 18:56

Contribution: 2014-12-10 18:57

Contribution: 2014-12-10 18:58

Contribution: 2014-12-11 18:54

Contribution: 2014-12-11 18:55

Contribution: 2014-12-11 18:56

Contribution: 2014-12-11 18:57

Contribution: 2014-12-11 18:58

Contribution: 2014-12-11 18:59

Contribution: 2014-12-12 18:54

Contribution: 2014-12-12 18:55

Contribution: 2014-12-12 18:56

Contribution: 2014-12-12 18:57

Contribution: 2014-12-12 18:58

Contribution: 2014-12-12 18:59

Contribution: 2014-12-18 18:54

Contribution: 2014-12-18 18:55

Contribution: 2014-12-18 18:56

Contribution: 2014-12-19 18:54

Contribution: 2014-12-19 18:55

Contribution: 2014-12-19 18:56

Contribution: 2014-12-19 18:57

Contribution: 2014-12-19 18:58

Contribution: 2014-12-22 18:54

Contribution: 2014-12-22 18:55

Contribution: 2014-12-22 18:56

Contribution: 2014-12-23 18:54

Contribution: 2014-12-23 18:55

Contribution: 2014-12-23 18:56

Contribution: 2014-12-24 18:54

Contribution: 2014-12-24 18:55

Contribution: 2014-12-24 18:56

Contribution: 2014-12-24 18:57

Contribution: 2014-12-24 18:58

Contribution: 2014-12-25 18:54

Contribution: 2014-12-25 18:55

Contribution: 2014-12-25 18:56

Contribution: 2014-12-25 18:57

Contribution: 2014-12-25 18:58

Contribution: 2014-12-30 18:54

Contribution: 2014-12-30 18:55

Contribution: 2014-12-31 18:54

Contribution: 2014-12-31 18:55

Contribution: 2014-12-31 18:56

Contribution: 2015-01-01 18:54

Contribution: 2015-01-02 18:54

Contribution: 2015-01-02 18:55

Contribution: 2015-01-02 18:56

Contribution: 2015-01-02 18:57

Contribution: 2015-01-05 18:54

Contribution: 2015-01-05 18:55

Contribution: 2015-01-05 18:56

Contribution: 2015-01-05 18:57

Contribution: 2015-01-05 18:58

Contribution: 2015-01-05 18:59

Contribution: 2015-01-06 18:54

Contribution: 2015-01-06 18:55

Contribution: 2015-01-06 18:56

Contribution: 2015-01-06 18:57

Contribution: 2015-01-07 18:54

Contribution: 2015-01-07 18:55

Contribution: 2015-01-08 18:54

Contribution: 2015-01-08 18:55

Contribution: 2015-01-08 18:56

Contribution: 2015-01-08 18:57

Contribution: 2015-01-09 18:54

Contribution: 2015-01-09 18:55

Contribution: 2015-01-09 18:56

Contribution: 2015-01-09 18:57

Contribution: 2015-01-09 18:58

Contribution: 2015-01-09 18:59

Contribution: 2015-01-09 19:00

Contribution: 2015-01-12 18:54

Contribution: 2015-01-12 18:55

Contribution: 2015-01-12 18:56

Contribution: 2015-01-12 18:57

Contribution: 2015-01-12 18:58

Contribution: 2015-01-12 18:59

Contribution: 2015-01-12 19:00

Contribution: 2015-01-13 18:54

Contribution: 2015-01-13 18:55

Contribution: 2015-01-13 18:56

Contribution: 2015-01-13 18:57



src/projects/Project.ts
export class Project {
  
  id: number | undefined;
  name: string = '';
  description: string = '';
  imageUrl: string = '';
  contractTypeId: number | undefined;
  contractSignedOn: Date = new Date();
  budget: number = 0;
  isActive: boolean = false;

  get isNew(): boolean {
    return this.id === undefined;
  }

  constructor(initializer?: any) {
    if (!initializer) return;
    if (initializer.id) this.id = initializer.id;
    if (initializer.name) this.name = initializer.name;
    if (initializer.description) this.description = initializer.description;
    if (initializer.imageUrl) this.imageUrl = initializer.imageUrl;
    if (initializer.contractTypeId)
      this.contractTypeId = initializer.contractTypeId;
    if (initializer.contractSignedOn)
      this.contractSignedOn = new Date(initializer.contractSignedOn);
    if (initializer.budget) this.budget = initializer.budget;
    if (initializer.isActive) this.isActive = initializer.isActive;
  }
}

src/projects/projectAPI.ts
import { Project } from './Project';
const baseUrl = 'http://localhost:4000';
const url = `${baseUrl}/projects`;

function translateStatusToErrorMessage(status: number) {
  switch (status) {
    case 401:
      return 'Please login again.';
    case 403:
      return 'You do not have permission to view the project(s).';
    default:
      return 'There was an error retrieving the project(s). Please try again.';
  }
}

function checkStatus(response: any) {
  if (response.ok) {
    return response;
  } else {
    const httpErrorInfo = {
      status: response.status,
      statusText: response.statusText,
      url: response.url,
    };
    console.log(`log server http error: ${JSON.stringify(httpErrorInfo)}`);

    let errorMessage = translateStatusToErrorMessage(httpErrorInfo.status);
    throw new Error(errorMessage);
  }
}

function parseJSON(response: Response) {
  return response.json();
}

// eslint-disable-next-line
function delay(ms: number) {
  return function (x: any): Promise<any> {
    return new Promise((resolve) => setTimeout(() => resolve(x), ms));
  };
}

function convertToProjectModels(data: any[]): Project[] {
  let projects: Project[] = data.map(convertToProjectModel);
  return projects;
}

function convertToProjectModel(item: any): Project {
  return new Project(item);
}

const projectAPI = {
  get(page = 1, limit = 15) {
    return fetch(`${url}?_page=${page}&_limit=${limit}&_sort=name`)
      .then(delay(600))
      .then(checkStatus)
      .then(parseJSON)
      .then(convertToProjectModels)
      .catch((error: TypeError) => {
        console.log('log client error ' + error);
        throw new Error(
          'There was an error retrieving the projects. Please try again.'
        );
      });
  },
  put(project: Project) {
      return fetch(`${url}/${project.id}`, {
        method: 'PUT',
        body: JSON.stringify(project),
        headers: {
          'Content-Type': 'application/json'
        }
      })
      .then(checkStatus)
      .then(parseJSON)
      .catch((error: TypeError) => {
        console.log('log client error ' + error);
        throw new Error(
          'There was an error updating the project. Please try again.'
        );
      });
  },
  find(id: number) {
    return fetch(`${url}/${id}`)
      .then(checkStatus)
      .then(parseJSON)
      .then(convertToProjectModel);
  },
};

export { projectAPI };

src/projects/MockProjects.ts
import { Project } from './Project';

export const MOCK_PROJECTS = [
  new Project({
    id: 1,
    name: 'Johnson - Kutch',
    description:
      'Fully-configurable intermediate framework. Ullam occaecati libero laudantium nihil voluptas omnis.',
    imageUrl: '/assets/placeimg_500_300_arch4.jpg',
    contractTypeId: 3,
    contractSignedOn: '2013-08-04T22:39:41.473Z',
    budget: 54637,
    isActive: false,
  }),
  new Project({
    id: 2,
    name: 'Wisozk Group',
    description:
      'Centralized interactive application. Exercitationem nulla ut ipsam vero quasi enim quos doloribus voluptatibus.',
    imageUrl: '/assets/placeimg_500_300_arch1.jpg',
    contractTypeId: 4,
    contractSignedOn: '2012-08-06T21:21:31.419Z',
    budget: 91638,
    isActive: true,
  }),
  new Project({
    id: 3,
    name: 'Denesik LLC',
    description:
      'Re-contextualized dynamic moratorium. Aut nulla soluta numquam qui dolor architecto et facere dolores.',
    imageUrl: '/assets/placeimg_500_300_arch12.jpg',
    contractTypeId: 6,
    contractSignedOn: '2016-06-26T18:24:01.706Z',
    budget: 29729,
    isActive: true,
  }),
  new Project({
    id: 4,
    name: 'Purdy, Keeling and Smitham',
    description:
      'Innovative 6th generation model. Perferendis libero qui iusto et ullam cum sint molestias vel.',
    imageUrl: '/assets/placeimg_500_300_arch5.jpg',
    contractTypeId: 4,
    contractSignedOn: '2013-05-26T01:10:42.344Z',
    budget: 45660,
    isActive: true,
  }),
  new Project({
    id: 5,
    name: 'Kreiger - Waelchi',
    description:
      'Managed logistical migration. Qui quod praesentium accusamus eos hic non error modi et.',
    imageUrl: '/assets/placeimg_500_300_arch12.jpg',
    contractTypeId: 2,
    contractSignedOn: '2009-12-18T21:46:47.944Z',
    budget: 81188,
    isActive: true,
  }),
  new Project({
    id: 6,
    name: 'Lesch - Waelchi',
    description:
      'Profound mobile project. Rem consequatur laborum explicabo sint odit et illo voluptas expedita.',
    imageUrl: '/assets/placeimg_500_300_arch1.jpg',
    contractTypeId: 3,
    contractSignedOn: '2016-09-23T21:27:25.035Z',
    budget: 53407,
    isActive: false,
  }),
];